compiler/rustc_lint/src/early.rs

```diff
@@ -275,28 +282,57 @@ impl<'a> ast_visit::Visitor<'a> for EarlyContextAndPasses<'a> {
     }
 
     fn visit_attribute(&mut self, attr: &'a ast::Attribute) {
-        run_early_passes!(self, check_attribute, attr);
+        lint_callback!(self, check_attribute, attr);
     }
 
     fn visit_mac_def(&mut self, mac: &'a ast::MacroDef, id: ast::NodeId) {
-        run_early_passes!(self, check_mac_def, mac);
+        lint_callback!(self, check_mac_def, mac);
         self.check_id(id);
     }
 
     fn visit_mac_call(&mut self, mac: &'a ast::MacCall) {
-        run_early_passes!(self, check_mac, mac);
+        lint_callback!(self, check_mac, mac);
         ast_visit::walk_mac(self, mac);
     }
 }
 
+// Combines multiple lint passes into a single pass, at runtime. Each
+// `check_foo` method in `$methods` within this pass simply calls `check_foo`
+// once per `$pass`. Compare with `declare_combined_early_lint_pass`, which is
+// similar, but combines lint passes at compile time.
+struct RuntimeCombinedEarlyLintPass<'a> {
+    passes: &'a mut [EarlyLintPassObject],
+}
```

Comment:
Instezad of having separate branches for the builtin-only vs builtin+others, can something like this work?
```suggestion
struct RuntimeCombinedEarlyLintPass<'a, T: EarlyLintPass> {
    builtin: T,
    passes: &'a mut [EarlyLintPassObject],
}
```

This would require changing the macros, but we may have both the static dispatch for builtin lints and limit the complexity.

Reply:
That won't help. Under the original structure, we had `builtin_lints` and `passes`, but `passes` is always for empty normal compilation. (It's non-empty when bootstrapping, because rustc has some rustc-specific lints, and it's also non-empty for clippy.)

I then then combined `builtin_lints` into `passes`. This meant that even in the normal builtins-only cases, the code had to iterate over the `passes` slice for every `check_*` call. This is what caused the slowdown. It took me a while to work out because the iterator methods are inlined so it wasn't that obvious.

With this new design, which is very similar to the original design, if `passes` is empty then we avoid all iteration over the slice. This gives the speed wins.

Reply:
Pseudocode may make it clearer. Simplifying greatly, it's basically the difference between this:
```
for node in ast {
    builtin_lints.check_node(node);
}
```
and this:
```
let passes = vec![builtin_lints]
for node in ast {
    for pass in passes {
        pass.check_node(node);
    }
}
```